IdentityCore/src/webview/response/MSIDWebOAuth2Response.m

```diff
@@ -26,8 +26,127 @@
+        response.url = url;
+        response.appInstallLink = appURLString;
+        
+        return response;
+    }
+    
+    // Check for auth response
+    // Try both the URL and the fragment parameters:
+    NSDictionary *parameters = [url msidFragmentParameters];
+    if (parameters.count == 0)
+    {
+        parameters = [url msidQueryParameters];
+    }
+    
+    // Verify state
+    BOOL stateVerified = stateVerifier(parameters, requestState);
+    
+    if (!stateVerified)
+    {
+        if (error){
+            *error = MSIDCreateError(MSIDOAuthErrorDomain, MSALErrorInvalidState, @"State returned from the server does not match", nil, nil, nil, context.correlationId, nil);
+            return nil;
+        }
+    }
+    
+    NSString *code = parameters[MSID_OAUTH2_CODE];
+    NSString *cloudHostName = parameters[MSID_AUTH_CLOUD_INSTANCE_HOST_NAME];
+    
+    if (code)
+    {
+        MSIDWebAADAuthResponse *response = [MSIDWebAADAuthResponse new];
+        response.url = url;
+        response.authorizationCode = code;
+        response.cloudHostName = cloudHostName;
+        return response;
+    }
+    
+    NSError *oauthError = [self.class oauthErrorFromDictionary:parameters];
+    if (!oauthError)
+    {
+        oauthError = MSIDCreateError(MSIDOAuthErrorDomain, MSIDErrorBadAuthorizationResponse, @"No code or error in server response.", nil, nil, nil, context.correlationId, nil);
+    }
+    
+    MSIDWebOAuth2Response *response = [MSIDWebOAuth2Response new];
+    response.url = url;
+    response.oauthError = oauthError;
+    
+    return response;
+    
+}
+
++ (NSError *)oauthErrorFromDictionary:(NSDictionary *)dictionary
+{
+    NSUUID *correlationId = [dictionary objectForKey:MSID_OAUTH2_CORRELATION_ID_RESPONSE] ?
+    [[NSUUID alloc] initWithUUIDString:[dictionary objectForKey:MSID_OAUTH2_CORRELATION_ID_RESPONSE]]:nil;
+    
+    NSString *serverOAuth2Error = [dictionary objectForKey:MSID_OAUTH2_ERROR];
+    
+    if (serverOAuth2Error)
+    {
+        NSString *errorDescription = dictionary[MSID_OAUTH2_ERROR_DESCRIPTION];
+        NSString *subError = dictionary[MSID_OAUTH2_CORRELATION_ID_RESPONSE];
+        
+        MSIDErrorCode errorCode = MSIDErrorCodeForOAuthError(errorDescription, MSIDErrorAuthorizationFailed);
+        
+        return MSIDCreateError(MSIDOAuthErrorDomain, errorCode, errorDescription, serverOAuth2Error, subError, nil, correlationId, nil);
+    }
+    
+    return nil;
+}
+
+
+// TODO: check if we have it in MSAL
++ (BOOL)verifyStateFromDictionary: (NSDictionary*) dictionary
+                          context:(id<MSIDRequestContext>)context
+{
+    NSDictionary *state = [NSDictionary msidURLFormDecode:[[dictionary objectForKey:MSID_OAUTH2_STATE] msidBase64UrlDecode]];
+    if (state.count != 0)
+    {
+        NSString *authorizationServer = [state objectForKey:@"a"];
+        NSString *resource            = [state objectForKey:@"r"];
```

Comment:
This looks very AAD v1 specific

Reply:
Yes, the state handling is also different.
Was manually merging JZ's code, this need to be changed;